### PR TITLE
Replace `kubectl apply` with `mesheryctl app onboard` command

### LINKED ISSUE #20: Use `mesheryctl` not bash
#### Current Behavior
Current tests use [bash scripts](https://github.com/layer5io/meshery-smp-action/tree/master/.github/workflows/scripts) to deploy service meshes and apps, but should be using `mesheryctl mesh` and `mesheryctl app`.

---
**[Contributor Guide](https://docs.meshery.io/project/contributing)**

### PR DESCRIPTION
**Pull request by alphaX86**
Signed-off-by: Aadhitya A <aadhitya864@gmail.com>

**Description**

This PR fixes #20 

**Notes for Reviewers**

**[Signed commits](../CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by leecalcote** [comment]
Does this behave as expected?

**Comment by alphaX86** [comment]
I thought it has CI actions inbuilt. Will check and let you know, Lee

**Comment by gyohuangxin** [comment]
@alphaX86 Hi, do you have any updates?

**Comment by alphaX86** [comment]
Heads Up! The adapter port nos are temporary for now... Plus, I need to test the scripts in a VM to ensure it works fine and safe

**Comment by alphaX86** [comment]
I just have only one tiny doubt. How you'd login to meshery via "None" provider, without the usage of auth token? It drives me crazy as the `--provider` flag is missing in mesheryctl

**Comment by leecalcote** [comment]
For Remote Provider, an infinite token can be assigned.

**Comment by alphaX86** [comment]
Well.. Its system login, because app on board needs a token to work. I introduced a provider flag and if its merged it'll be good to go for this script

**Comment by alphaX86** [comment]
> @alphaX86 Thanks.
> 
> > some PRs need to be merged on Meshery repo in order to reflect here
> 
> So `mesheryctl app onboard` command cannot work at this stage, right?
> 
> > @gyohuangxin any suggestions to test the shell script on a remote cloud system with k8s running. Mine has reached limits due to t2.micro CPU
> 
> The fastest way is to use this github action on your forked repo, you need to set `enhance` branch as the default branch, then you can enable action workflows in https://github.com/alphaX86/meshery-smp-action/actions. Then run configurable-benchmark-test, it will deploy k8s and run your changed scripts to deploy meshes and applications. This way doesn't need any other resource. image

I'll check it out. Thank you so much

**Comment by gyohuangxin** [comment]
> Well.. Its system login, because `app onboard` needs a token to work. I introduced a provider flag and if its merged it'll be good to go for this script

Is it there any workaround to let it work now? Because I will work on enabling other meshes and try to use `mesheryctl` to deploy meshes and apps these days.

**Comment by alphaX86** [comment]
Ok, you'd need an infinite lifetime auth token to accomplish this, because `system login` is prompt based for now and it won't work on pure bash script. Though 'none' provider can be logged in, I had deep trouble to access via script which eventually failed on all methods

The token alone however, some folks only have it... @Revolyssup anyway to acquire this token?

**Comment by gyohuangxin** [comment]
@alphaX86 Thanks for the information 👍 I have one infinite lifetime auth token in fact.

**Comment by alphaX86** [comment]
Then it'd be easy for you 🙂👍 thanks @gyohuangxin

File: .github/workflows/scripts/istio_deploy.sh [hunk 1]
```diff
@@ -1,21 +1,23 @@
...
 
-curl -L https://istio.io/downloadIstio | ISTIO_VERSION=1.11.4 sh -
-cd istio-1.11.4
-export PATH=$PWD/bin:$PATH
-istioctl install --set profile=demo -y
-kubectl label namespace default istio-injection=enabled
+# Check if mesheryctl is present, else install it
+if ! [ -x "$(command -v mesheryctl)" ]; then
+    echo 'mesheryctl is not installed. Installing mesheryctl client... Standby... (Starting Meshery as well...)' >&2
+    curl -L https://meshery.io/install | PLATFORM=kubernetes bash -
```
Comment (gyohuangxin) [inline .github/workflows/scripts/istio_deploy.sh:13]: We don't need deploy all meshes adapters.
```suggestion
    curl -L https://meshery.io/install | ADAPTERS=istio PLATFORM=kubernetes bash -
```

File: .github/workflows/scripts/linkerd_deploy.sh [hunk 2]
```diff
@@ -8,16 +8,18 @@
...
-curl -fsL https://run.linkerd.io/install | sh
-export PATH=$PATH:/home/runner/.linkerd2/bin
-linkerd version
-linkerd check --pre
-linkerd install | kubectl apply -f -
-linkerd check
+# Check if mesheryctl is present, else install it
+if ! [ -x "$(command -v mesheryctl)" ]; then
+    echo 'mesheryctl is not installed. Installing mesheryctl client... Standby... (Starting Meshery as well...)' >&2
+    curl -L https://meshery.io/install | PLATFORM=kubernetes bash -
```
Comment (gyohuangxin) [inline .github/workflows/scripts/linkerd_deploy.sh:14]: Ditto.

File: .github/workflows/scripts/osm_deploy.sh [hunk 3]
```diff
@@ -8,18 +8,17 @@
...
-    --set=OpenServiceMesh.deployGrafana=true \
-    --set=OpenServiceMesh.deployJaeger=true
-
-kubectl create namespace bookstore
-./${system,,}-amd64/osm namespace add bookstore
-kubectl apply -f https://raw.githubusercontent.com/openservicemesh/osm/release-v0.11/docs/example/manifests/apps/bookstore.yaml
+# Check if mesheryctl is present, else install it
+if ! [ -x "$(command -v mesheryctl)" ]; then
+    echo 'mesheryctl is not installed. Installing mesheryctl client... Standby... (Starting Meshery as well...)' >&2
+    curl -L https://meshery.io/install  | PLATFORM=kubernetes bash -
```
Comment (gyohuangxin) [inline .github/workflows/scripts/osm_deploy.sh:14]: Ditto

File: .github/workflows/scripts/linkerd_deploy.sh (hunk 2 above)
Comment (leecalcote) [reply .github/workflows/scripts/linkerd_deploy.sh:14]: Please do suggest and let's get the change committed. `ADAPTERS=linkerd`?

**Comment by alphaX86** [comment]
A release of Meshery is needed to reflect the changes of `system login` (PR merged but it's not made public yet). Tests are run, so far good. If a release alone is made in Meshery repo (I'm not forcing it to so, just informing) then changes will reflect on script files as well. Other than login, everything worked pretty fine (except `mesh` and `app` due to login issue)

**Comment by gyohuangxin** [comment]
@alphaX86 Great! Thanks for your information. 
A question not related to this PR, can we use `mesheryctl app aboard`  to deploy some demo apps for Kuma, Consul, Appmesh or Cilium? I'm onboarding them into GH action, but confused about how to deploy demo apps on them.

**Comment by alphaX86** [comment]
`mesheryctl app onboard` is generic command and should work on all service meshes. To onboard an app, you'd need a YAML file containing the properties of app to be onboarded or a GitHub link referencing the file. I still didn't test the command fully on action but will test on my system and let you know further

**Comment by gyohuangxin** [comment]
Thanks. It would be great if we can aboard any app on any mesh.

File: .github/workflows/scripts/istio_deploy.sh [hunk 4]
```diff
@@ -1,21 +1,23 @@
...
+if ! [ -x "$(command -v mesheryctl)" ]; then
+    echo 'mesheryctl is not installed. Installing mesheryctl client... Standby... (Starting Meshery as well...)' >&2
+    curl -L https://meshery.io/install | ADAPTERS=istio PLATFORM=kubernetes bash -
+fi
 
-kubectl apply -f samples/bookinfo/platform/kube/bookinfo.yaml
-kubectl apply -f samples/bookinfo/networking/bookinfo-gateway.yaml
+sleep 10
+#mesheryctl system login --provider None
+echo 'E' | mesheryctl mesh deploy adapter meshery-istio:10000 --token "./.github/workflows/auth.json"
```
Comment (gyohuangxin) [inline .github/workflows/scripts/istio_deploy.sh:18]: Probably the Istio didn't deploy completely, can we add a `sleep` and a check here?

**Review by gyohuangxin — COMMENTED** [review]
@alphaX86 Does it only happen on Istio? It seems Istio didn't deployed properly.

**Comment by alphaX86** [comment]
Sure 👍 I'll add and check

File: .github/workflows/scripts/istio_deploy.sh (hunk 4 above)
Comment (alphaX86) [reply .github/workflows/scripts/istio_deploy.sh:18]: Istio is deployed, I found a possible cause now, size of minikube cluster isn't sufficient, I'll change size and check

File: .github/workflows/configurable-benchmark-test.yaml [hunk 5]
```diff
@@ -35,12 +35,13 @@ jobs:
     if: ${{ github.event_name == 'workflow_dispatch' }}
     steps:
       - name: Setup Kubernetes
-        uses: manusa/actions-setup-minikube@v2.4.3
+        uses: medyagh/setup-minikube@v0.0.7
```
Comment (gyohuangxin) [inline .github/workflows/configurable-benchmark-test.yaml:38]: You can set cpu and memory like: `start args: "--cpu 4 --memory 5192"` without changing repo.

**Comment by gyohuangxin** [comment]
@alphaX86 Do you have any updates?

**Comment by alphaX86** [comment]
@gyohuangxin nothing solid, the k8s deprecation error was found to occur on old k8s resources found on manifests, needs changes on that side, then only `pattern apply` would work flawless. Other issue like error 500 is common in pattern it seems after talks last week

**Comment by alphaX86** [comment]
> > the k8s deprecation error was found to occur on old k8s resources found on manifests
> 
> I'm confused about it. Can you show the error log? Does it occur after using medyagh/setup-minikube@v0.0.7?

It was produced earlier and it's disappearing somehow (I'm also confused). And no, it's not occuring due to new setup-minikube action

**Comment by gyohuangxin** [comment]
@hershd23 Can you help try this?

**Comment by gyohuangxin** [comment]
@hershd23 Can you help try this?

**Comment by hershd23** [comment]
Yep I'll give this a shot

**Comment by hershd23** [comment]
https://github.com/hershd23/meshery-smp-action/runs/6867611798?check_suite_focus=true#step:5:226

Currently the tmate step keeps on printing the link to connect to the common terminal. It doesn't reach the next task. Any fixes for this @alphaX86 ?

**Comment by hershd23** [comment]
https://github.com/hershd23/meshery-smp-action/actions/runs/2490527037

@gyohuangxin @alphaX86 still getting familiar with the issue, pulled up @alphaX86 's latest changes and the `pattern apply` works fine. Checking for `app onboard`

File: .github/workflows/scripts/linkerd_deploy.sh [hunk 6]
```diff
@@ -8,16 +8,18 @@
...
+# Check if mesheryctl is present, else install it
+if ! [ -x "$(command -v mesheryctl)" ]; then
+    echo 'mesheryctl is not installed. Installing mesheryctl client... Standby... (Starting Meshery as well...)' >&2
+    curl -L https://meshery.io/install | ADAPTERS=linkerd PLATFORM=kubernetes bash -
+fi
 
-curl -fsL https://run.linkerd.io/emojivoto.yml | kubectl apply -f -
-kubectl -n emojivoto port-forward svc/web-svc 8080:80 &> /dev/null &
-kubectl get -n emojivoto deploy -o yaml | linkerd inject - | kubectl apply -f -
+curl -fsL https://run.linkerd.io/emojivoto.yml
```
Comment (hershd23) [inline .github/workflows/scripts/linkerd_deploy.sh:17]: The curl command isn't putting the output in a file. For reference :- https://github.com/hershd23/meshery-smp-action/runs/6868074975?check_suite_focus=true#step:4:285

We can add it explicitly like shown

```suggestion
curl -fsL https://run.linkerd.io/emojivoto.yml --output emojivoto.yml
```

**Comment by hershd23** [comment]
https://github.com/hershd23/meshery-smp-action/runs/6868235120?check_suite_focus=true

Linkerd + wrk2 combination working fine with `app onboard`

Are there specific cases that are not working @gyohuangxin @alphaX86 @leecalcote, that I should test?

**Comment by alphaX86** [comment]
> https://github.com/hershd23/meshery-smp-action/runs/6867611798?check_suite_focus=true#step:5:226
> 
> Currently the tmate step keeps on printing the link to connect to the common terminal. It doesn't reach the next task. Any fixes for this @alphaX86 ?
> 
> Edit :- Found a way by creating a file named continue in the project directory, is there a better way?

Tmate is used to debug the current session actually @hershd23 that is, you need to open the link and type :wq then the shell will be activated. Once you finished debugging you need to exit the shell by exit command. Then only it'll proceed to next step of action

**Comment by gyohuangxin** [comment]
@hershd23 Thanks for this testing, I saw the `500` error message in your testing of `mesheryctl pattern apply` on Istio: https://github.com/hershd23/meshery-smp-action/runs/6867816269?check_suite_focus=true#step:4:69. Did you check whether the pattern is working well?

**Comment by gyohuangxin** [comment]
> https://github.com/hershd23/meshery-smp-action/runs/6868235120?check_suite_focus=true
> 
> Linkerd + wrk2 combination working fine with app onboard
> 
> Are there specific cases that are not working @gyohuangxin @alphaX86 @leecalcote, that I should test?

IOM, the `app onboard` or `pattern apply` can achieve our goal. We prefer `pattern apply` because we can implement more interesting things on it in the future. Therefore, we should make sure at this time:
1. We can use `mesheryctl` to deploy each mesh (Istio, Linkerd, OSM) in github action. Make sure this command works well: https://github.com/layer5io/meshery-smp-action/pull/48/files#diff-9708a03424bc77988703c3ef86daa584385ea618a3d48e55842e537d22620444R14
2. We can use `pattern apply` to deploy `pattern` on each mesh and the performance tests works well.

**Comment by hershd23** [comment]
@gyohuangxin you're right

Basically the pattern apply command is erroring out, the trace isn't helpful

```
runner@fv-az204-655:~/work/meshery-smp-action/meshery-smp-action$ mesheryctl pattern apply -f "https://github.com/service-mesh-patterns/service-mesh-patterns/blob/master/samples/bookInfoPattern.yaml" --token "./.github/workflows/auth.json"
Error: Response Status Code 500, possible Server Error
```

**Comment by hershd23** [comment]
```
meshery-7bfb4d5c9c-dzplb        |       time="2022-06-14T18:10:00Z" level=error msg="readdirent /tmp/service-mesh-patterns/1655230199991602250/samples/bookInfoPattern.yaml: not a directory" app=meshery code=2167 probable-cause="Cannot save the Pattern due to wrong path or URL" severity=2 short-description="Error failed to import pattern" suggested-remediation="Check if the given path or URL of the Pattern is correct"
meshery-7bfb4d5c9c-dzplb        |       time="2022-06-14T18:12:52Z" level=error msg="readdirent /tmp/service-mesh-patterns/1655230372443385636/samples/bookInfoPattern.yaml: not a directory" app=meshery code=2167 probable-cause="Cannot save the Pattern due to wrong path or URL" severity=2 short-description="Error failed to import pattern" suggested-remediation="Check if the given path or URL of the Pattern is correct"
meshery-7bfb4d5c9c-dzplb        |       time="2022-06-14T18:12:55Z" level=error msg="readdirent /tmp/service-mesh-patterns/1655230375138899338/samples/bookInfoPattern.yaml: not a directory" app=meshery code=2167 probable-cause="Cannot save the Pattern due to wrong path or URL" severity=2 short-description="Error failed to import pattern" suggested-remediation="Check if the given path or URL of the Pattern is correct"
```

Checked the system logs got this

**Comment by hershd23** [comment]
```
runner@fv-az204-655:~/work/meshery-smp-action/meshery-smp-action$ mesheryctl pattern apply -f "https://raw.githubusercontent.com/service-mesh-patterns/service-mesh-patterns/master/samples/bookInfoPattern.yaml" --token "./.github/workflows/auth.json"
pattern successfully applied
successfully deployed application productpage
successfully deployed application reviews-v1
successfully deployed application reviews-v3
successfully deployed application details
successfully deployed application ratings
created Gateway "sample-app-gateway" in namespace "default"
applied trait "meshmap" on service "reviews-v2"
applied trait "meshmap" on service "reviews-v3"
applied trait "meshmap" on service "reviews"
applied trait "meshmap" on service "reviews-v1"
applied trait "meshmap" on service "productpage"
applied trait "meshmap" on service "ratings"
applied trait "meshmap" on service "sample-app-gateway"
applied trait "meshmap" on service "bookinfo-vs"
applied trait "meshmap" on service "details"
created VirtualService "bookinfo-vs" in namespace "default"
applied trait "meshmap" on service "reviews-v2"
applied trait "meshmap" on service "reviews-v3"
applied trait "meshmap" on service "reviews"
applied trait "meshmap" on service "reviews-v1"
applied trait "meshmap" on service "productpage"
applied trait "meshmap" on service "ratings"
applied trait "meshmap" on service "sample-app-gateway"
applied trait "meshmap" on service "bookinfo-vs"
applied trait "meshmap" on service "details"
successfully deployed application reviews-v2
runner@fv-az204-655:~/work/meshery-smp-action/meshery-smp-action$ kubectl get pods
NAME                           READY   STATUS              RESTARTS   AGE
details-69648f48b4-rc8qf       0/1     ContainerCreating   0          4s
productpage-76dc999cf7-4j4d6   0/1     ContainerCreating   0          4s
ratings-6dfcc745f-drn9k        0/1     ContainerCreating   0          4s
reviews-v1-dbcdfd648-9shdd     0/1     ContainerCreating   0          4s
reviews-v2-7bb7dd9fc-jfnjz     0/1     ContainerCreating   0          4s
reviews-v3-784c4996f9-4rqhm    0/1     ContainerCreating   0          4s
```

Ah simply changing to download from raw file works!

https://raw.githubusercontent.com/service-mesh-patterns/service-mesh-patterns/master/samples/bookInfoPattern.yaml

**Comment by hershd23** [comment]
The run is successful

It has an error towards the end, I don't know how relevant that is

https://github.com/hershd23/meshery-smp-action/runs/6886559987?check_suite_focus=true#step:6:1709

```
Error getting context: json: cannot unmarshal array into Go value of type map[string]interface {}
```
Any ideas @gyohuangxin @alphaX86 @leecalcote ?

**Comment by hershd23** [comment]
Another comment, I feel for `mesheryctl pattern apply` one of the most basic errors would be that the URL is wrong or the file cannot be downloaded from that URL. We should have a better error message :)

```
runner@fv-az204-655:~/work/meshery-smp-action/meshery-smp-action$ mesheryctl pattern apply -f "https://github.com/service-mesh-patterns/service-mesh-patterns/blob/master/samples/bookInfoPattern.yaml" --token "./.github/workflows/auth.json"
Error: Response Status Code 500, possible Server Error
```

**Comment by hershd23** [comment]
@gyohuangxin should I also change the linkerd task from `app onboard` to `pattern apply` and test?

Or let it be as is?

**Comment by gyohuangxin** [comment]
> @gyohuangxin should I also change the linkerd task from `app onboard` to `pattern apply` and test?
> 
> Or let it be as is?

Yes, we should keep the method of deployment the same and also test on OSM.

**Comment by gyohuangxin** [comment]
> Also should I be making my own PR on this issue? Or suggest changes to this one itself? Not sure of @alphaX86 's availability due to his exams

Your own PR makes sense.

**Comment by alphaX86** [comment]
You're cleared to make a PR @hershd23 I'll close this staled one

**Comment by gyohuangxin** [comment]
@alphaX86 Thank you very much!

**Comment by alphaX86** [comment]
Strange.... I sent the above mail comment 2-3 months ago. Seems a bug within GitHub :/ anyway the changes are already merged in other way

**Comment by gyohuangxin** [comment]
@alphaX86 OK fine, but the scripts doesn't work indeed recently. We tried to debug but have no idea, https://layer5io.slack.com/archives/CVBGPTFJN/p1665299033777169?thread_ts=1664371772.250329&cid=CVBGPTFJN, can you have a look if you have some cycles

**Comment by alphaX86** [comment]
Got it

**Comment by gyohuangxin** [comment]
> Got it 👍

Thank you

**Pull request closed without merging**